src/rideables/Stack.hpp

```diff
@@ -0,0 +1,104 @@
+// #ifndef STACK_HPP
+// #define STACK_HPP
+
+#include <stdio.h>
+#include <iostream>
+#include <atomic>
+#include <algorithm>
+#include "HarnessUtils.hpp"
+#include "ConcurrentPrimitives.hpp"
+#include "RQueue.hpp"
+#include "RCUTracker.hpp"
+#include "CustomTypes.hpp"
+
+using namespace std;
+
+template <typename T>
+class Stack
+{
+
+private:
+    struct StackNode
+    {
+        int data;
+        StackNode *next;
+    };
+
+    atomic<StackNode *> top;
+    RCUTracker<StackNode> tracker;
+
+public:
+    Stack(int task_num) : top(nullptr), tracker(task_num, 100, 1000, true) {}
+    void push(int data, int tid);
+    optional<T> pop(int tid);
+};
+
+template <typename T>
+void Stack<T>::push(int data, int tid)
+{
+    tracker.start_op(tid);
+
+    StackNode *new_node = new StackNode;
+    new_node->data = data;
+    new_node->next = NULL;
+
+    StackNode *old_node;
+
+    do
+    {
+        old_node = top.load();
+        new_node->next = old_node;
+    } while (!top.compare_exchange_weak(old_node, new_node));
+
+    tracker.end_op(tid);
+}
+
+template <typename T>
+optional<T> Stack<T>::pop(int tid)
+{
+    tracker.start_op(tid);
+    StackNode *new_node;
+    StackNode *old_node;
+    optional<T> res = {};
+    do
+    {
+        old_node = top.load();
+        if (old_node == NULL)
+        {
+            tracker.end_op(tid);
+            res = "Error : popping an empty stack";
```

Comment:
returning `res` without assigning any other value to it would be alright; it's initialized to empty at line 62

Reply:
also, assigning a string to optional<T> assumes that T is std::string, which sometimes is wrong